File: python-impl/Lexer.py
from Token import Token, TokenType
from Utils import Assert

keywords: dict = {
    "if": TokenType.IF,
    "else": TokenType.ELSE,
    "true": TokenType.TRUE,
    "false": TokenType.FALSE,
    "nil": TokenType.NIL,
    "while": TokenType.WHILE,
    "function": TokenType.FUNCTION,
    "return": TokenType.RETURN,
    "and": TokenType.AND,
    "or": TokenType.OR,
    "not": TokenType.NOT,
    "struct": TokenType.STRUCT,
    "ref": TokenType.REF,
    "dllimport": TokenType.DLLIMPORT,
    "import": TokenType.IMPORT,
}


def IsNumber(c) -> bool:
    return c >= '0' and c <= '9'


def IsLetter(c) -> bool:
    return (c >= 'A' and c <= 'Z') or (c >= 'a' and c <= 'z') or c == '_'


def IsLetterOrNumber(c) -> bool:
    return IsLetter(c) or IsNumber(c)


class Lexer:
    __startPos: int = 0
    __curPos: int = 0
    __line: int = 1
    __source: str = ""
    __tokens: list = []
    __filePath: str = ""

    def __init__(self) -> None:
        self.ResetStatus()

    def GenerateTokens(self, src: str, filePath="RootFile") -> list:
        self.ResetStatus()
        self.__source = src
        self.__filePath = filePath
        while (not self.IsAtEnd()):
            self.__startPos = self.__curPos
            self.GenerateToken()

        return self.__tokens

    def ResetStatus(self) -> None:
        self.__startPos = self.__curPos = 0
        self.__line = 1
        self.__tokens = []
        self.__source = ""

    def IsMatchCurChar(self, c) -> bool:
        return self.GetCurChar() == c

    def IsMatchCurCharAndStepOnce(self, c) -> bool:
        result = self.GetCurChar() == c
        if result:
            self.__curPos += 1
        return result

    def IsMatchNextChar(self, c) -> bool:
        return self.GetNextChar() == c

    def IsMatchNextCharAndStepOnce(self, c) -> bool:
        result = self.GetNextChar() == c
        if result:
            self.__curPos += 1
        return result

    def GetNextCharAndStepOnce(self):
        if self.__curPos+1 < len(self.__source):
            self.__curPos += 1
            return self.__source[self.__curPos]
        return '\0'

    def GetNextChar(self):
        if self.__curPos+1 < len(self.__source):
            return self.__source[self.__curPos+1]
        return '\0'

    def GetCurCharAndStepOnce(self):
        res = '\0'
        if not self.IsAtEnd():
            res = self.__source[self.__curPos]
            self.__curPos += 1
        return res

    def GetCurChar(self):
        res = '\0'
        if not self.IsAtEnd():
            res = self.__source[self.__curPos]
        return res

    def AddToken(self, type: TokenType, literal: str = None):
        if (literal == None):
            literal = self.__source[self.__startPos:self.__curPos]
        self.__tokens.append(
            Token(type, literal, self.__line, self.__filePath))

    def IsAtEnd(self) -> bool:
        return self.__curPos >= len(self.__source)

    def Number(self):
        while IsNumber(self.GetCurChar()):
            self.GetCurCharAndStepOnce()
        if self.IsMatchCurCharAndStepOnce('.'):
            if IsNumber(self.GetCurChar()):
                while IsNumber(self.GetCurChar()):
                    self.GetCurCharAndStepOnce()
            else:
                Assert("[line " + self.__line + "]:Number cannot end with '.'")
        self.AddToken(TokenType.NUMBER)

    def Identifier(self):
        while IsLetterOrNumber(self.GetCurChar()):
            self.GetCurCharAndStepOnce()

        literal = self.__source[self.__startPos:self.__curPos]

        isKeyWord = False
        for key, value in keywords.items():
            if key == literal:
                self.AddToken(value, literal)
                isKeyWord = True
                break
        if not isKeyWord:
            self.AddToken(TokenType.IDENTIFIER, literal)

    def String(self):
        while (not self.IsMatchCurChar('\"')) and (not self.IsAtEnd()):
            if self.IsMatchCurChar('\n'):
                self.__line += 1
            self.GetCurCharAndStepOnce()
        if self.IsAtEnd():
            Assert("[line " + self.__line + "]:Uniterminated string.")
        self.GetCurCharAndStepOnce()  # eat the second '\"'

        self.AddToken(TokenType.STRING,
                      self.__source[self.__startPos+1:self.__curPos-1])

    def GenerateToken(self):
        c = self.GetCurCharAndStepOnce()

        if c == '(':
            self.AddToken(TokenType.LPAREN)
        elif c == ')':
            self.AddToken(TokenType.RPAREN)
        elif c == '[':
            self.AddToken(TokenType.LBRACKET)
        elif c == ']':
            self.AddToken(TokenType.RBRACKET)
        elif c == '{':
            self.AddToken(TokenType.LBRACE)
        elif c == '}':
            self.AddToken(TokenType.RBRACE)
        elif c == ',':
            self.AddToken(TokenType.COMMA)
        elif c == '.':
            self.AddToken(TokenType.DOT)
        elif c == ':':
            self.AddToken(TokenType.COLON)
        elif c == ';':
            self.AddToken(TokenType.SEMICOLON)
        elif c == '\"':
            self.String()
        elif c == ' ' or c == ' \t' or c == '\r':
            pass
        elif c == '\n':
            self.__line += 1
        elif c == '+':
            self.AddToken(TokenType.PLUS)
        elif c == '-':
            self.AddToken(TokenType.MINUS)
        elif c == '*':
            self.AddToken(TokenType.ASTERISK)
        elif c == '/':
            self.AddToken(TokenType.SLASH)
        elif c == '#':
            while (not self.IsMatchCurChar('\n')) and (not self.IsAtEnd()):
                self.GetCurCharAndStepOnce()
            self.__line += 1
        elif c == '!':
            if self.IsMatchCurCharAndStepOnce('='):
                self.AddToken(TokenType.BANG_EQUAL)
        elif c == '<':
            if self.IsMatchCurCharAndStepOnce('='):
                self.AddToken(TokenType.LESS_EQUAL)
            else:
                self.AddToken(TokenType.LESS)
        elif c == '>':
            if self.IsMatchCurCharAndStepOnce('='):
                self.AddToken(TokenType.GREATER_EQUAL)
            else:
                self.AddToken(TokenType.GREATER)
        elif c == '=':
            if self.IsMatchCurCharAndStepOnce('='):
                self.AddToken(TokenType.EQUAL_EQUAL)
            else:
                self.AddToken(TokenType.EQUAL)
        elif IsNumber(c):
            self.Number()
        elif IsLetter(c):
            self.Identifier()
        else:
            self.AddToken(TokenType.UNKNOWN)

File: otherImpl/c#/libComputeDuck/Lexer.cs
using System.Collections.Generic;

namespace ComputeDuck
{
    public class Lexer
    {
        public Lexer()
        {
            ResetStatus();
        }

        public List<Token> GenerateTokens(string src, string filePath = "RootFile")
        {
            ResetStatus();
            m_Source = src;
            m_FilePath = filePath;
            while (!IsAtEnd())
            {
                m_StartPos = m_CurPos;
                GenerateToken();
            }

            return m_Tokens;
        }

        private void ResetStatus()
        {
            m_StartPos = m_CurPos = 0;
            m_Line = 1;
            m_Column = 1;
            m_Tokens = new List<Token>();
        }


        private void GenerateToken()
        {
            char c = GetCurCharAndStepOnce();
            switch (c)
            {
                case '(':
                    AddToken(TokenType.LPAREN);
                    break;
                case ')':
                    AddToken(TokenType.RPAREN);
                    break;
                case '[':
                    AddToken(TokenType.LBRACKET);
                    break;
                case ']':
                    AddToken(TokenType.RBRACKET);
                    break;
                case '{':
                    AddToken(TokenType.LBRACE);
                    break;
                case '}':
                    AddToken(TokenType.RBRACE);
                    break;
                case ',':
                    AddToken(TokenType.COMMA);
                    break;
                case '.':
                    AddToken(TokenType.DOT);
                    break;
                case ':':
                    AddToken(TokenType.COLON);
                    break;
                case ';':
                    AddToken(TokenType.SEMICOLON);
                    break;
                case '\"':
                    String();
                    break;
                case ' ':
                case '\t':
                case '\r':
                    break;
                case '\n':
                    m_Line++;
                    m_Column = 1;
                    break;
                case '+':
                    AddToken(TokenType.PLUS);
                    break;
                case '-':
                    AddToken(TokenType.MINUS);
                    break;
                case '*':
                    AddToken(TokenType.ASTERISK);
                    break;
                case '/':
                    AddToken(TokenType.SLASH);
                    break;
                case '&':
                    AddToken(TokenType.AMPERSAND);
                    break;
                case '|':
                    AddToken(TokenType.VBAR);
                    break;
                case '~':
                    AddToken(TokenType.TILDE);
                    break;
                case '^':
                    AddToken(TokenType.CARET);
                    break;
                case '#':
                    {
                        while (!IsMatchCurChar('\n') && !IsAtEnd())
                            GetCurCharAndStepOnce();
                        break;
                    }
                case '!':
                    if (IsMatchCurCharAndStepOnce('='))
                        AddToken(TokenType.BANG_EQUAL);
                    break;
                case '<':
                    if (IsMatchCurCharAndStepOnce('='))
                        AddToken(TokenType.LESS_EQUAL);
                    else
                        AddToken(TokenType.LESS);
                    break;
                case '>':
                    if (IsMatchCurCharAndStepOnce('='))
                        AddToken(TokenType.GREATER_EQUAL);
                    else
                        AddToken(TokenType.GREATER);
                    break;
                case '=':
                    if (IsMatchCurCharAndStepOnce('='))
                        AddToken(TokenType.EQUAL_EQUAL);
                    else
                        AddToken(TokenType.EQUAL);
                    break;
                default:
                    if (IsNumber(c))
                        Number();
                    else if (IsLetter(c))
                        Identifier();
                    else
                        AddToken(TokenType.END);
                    break;
            }
        }

        private bool IsMatchCurChar(char c)
        {
            return GetCurChar() == c;
        }

        private bool IsMatchCurCharAndStepOnce(char c)
        {
            bool result = GetCurChar() == c;
            if (result)
            {
                m_CurPos++;
                m_Column++;
            }
            return result;
        }

        private char GetCurCharAndStepOnce()
        {
            if (!IsAtEnd())
            {
                m_Column++;
                return m_Source[m_CurPos++];
            }
            return '\0';
        }
        private char GetCurChar()
        {
            if (!IsAtEnd())
                return m_Source[m_CurPos];
            return '\0';
        }

        private void AddToken(TokenType type)
        {
            var literal = m_Source.Substring(m_StartPos, m_CurPos - m_StartPos);
            m_Tokens.Add(new Token(type, literal, m_Line,m_Column, m_FilePath));
        }
        private void AddToken(TokenType type, string literal)
        {
            m_Tokens.Add(new Token(type, literal, m_Line, m_Column, m_FilePath));
        }

        private bool IsAtEnd()
        {
            return m_CurPos >= m_Source.Length;
        }

        bool IsNumber(char c)
        {
            return c >= '0' && c <= '9';
        }
        bool IsLetter(char c)
        {
            return (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || c == '_';
        }
        bool IsLetterOrNumber(char c)
        {
            return IsLetter(c) || IsNumber(c);
        }

        void Number()
        {
            while (IsNumber(GetCurChar()))
                GetCurCharAndStepOnce();

            if (IsMatchCurCharAndStepOnce('.'))
            {
                if (IsNumber(GetCurChar()))
                    while (IsNumber(GetCurChar()))
                        GetCurCharAndStepOnce();
                else
                    Utils.Assert("[line " + m_Line.ToString() + "]:Number cannot end with '.'");
            }
            AddToken(TokenType.NUMBER);
        }
        void Identifier()
        {
            while (IsLetterOrNumber(GetCurChar()))
                GetCurCharAndStepOnce();

            string literal = m_Source.Substring(m_StartPos, m_CurPos - m_StartPos);

            bool isKeyWord = false;
            foreach (var entry in keywords)
                if (entry.Key == literal)
                {
                    AddToken(entry.Value, literal);
                    isKeyWord = true;
                    break;
                }

            if (!isKeyWord)
                AddToken(TokenType.IDENTIFIER, literal);
        }
        void String()
        {
            while (!IsMatchCurChar('\"') && !IsAtEnd())
            {
                if (IsMatchCurChar('\n'))
                    m_Line++;
                GetCurCharAndStepOnce();
            }

            if (IsAtEnd())
                Utils.Assert("[line " + m_Line + "]:Uniterminated string.");

            GetCurCharAndStepOnce(); //eat the second '\"'

            AddToken(TokenType.STRING, m_Source.Substring(m_StartPos + 1, m_CurPos - m_StartPos - 2));
        }

        private int m_StartPos;
        private int m_CurPos;
        private int m_Line;
        private int m_Column;
        private string m_Source;
        private List<Token> m_Tokens;

        private string m_FilePath;

        private static Dictionary<string, TokenType> keywords = new Dictionary<string, TokenType>
        {
            {"if", TokenType.IF},
            {"else", TokenType.ELSE},
            {"true", TokenType.TRUE},
            {"false", TokenType.FALSE},
            {"nil", TokenType.NIL},
            {"while", TokenType.WHILE},
            {"function", TokenType.FUNCTION},
            {"return", TokenType.RETURN},
            {"and", TokenType.AND},
            {"or", TokenType.OR},
            {"not", TokenType.NOT},
            {"struct", TokenType.STRUCT},
            {"ref", TokenType.REF},
            {"dllimport",TokenType.DLLIMPORT},
            {"import",TokenType.IMPORT},
        };
    }
}